File: fixtures/scale/dart/medium/lib/parser/token_cursor.dart
import '../errors/parse_exception.dart';
import '../lexer/token.dart';
import '../lexer/token_type.dart';

class TokenCursor {
  final List<Token> tokens;
  int _index = 0;

  TokenCursor(this.tokens);

  Token peek() => tokens[_index];

  Token advance() => tokens[_index++];

  bool atType(TokenType type) => peek().type == type;

  bool accept(TokenType type) {
    if (!atType(type)) return false;
    _index++;
    return true;
  }

  void expect(TokenType type) {
    if (advance().type != type) throw ParseException('expected $type');
  }
}

File: fixtures/scale/java/medium/src/calc/parser/TokenCursor.java
package calc.parser;

import java.util.List;

import calc.errors.ParseException;
import calc.lexer.Token;
import calc.lexer.TokenType;

public class TokenCursor {
    private final List<Token> tokens;
    private int index;

    public TokenCursor(List<Token> tokens) {
        this.tokens = tokens;
    }

    public Token peek() { return tokens.get(index); }

    public Token advance() { return tokens.get(index++); }

    public boolean atType(TokenType type) { return peek().getType() == type; }

    public boolean accept(TokenType type) {
        if (!atType(type)) return false;
        index++;
        return true;
    }

    public void expect(TokenType type) {
        if (advance().getType() != type) {
            throw new ParseException("expected " + type);
        }
    }
}
